docs/proposals/pvcs-smaller-than-1gi.md

```diff
@@ -0,0 +1,543 @@
+        }
+        // ...
+        var lv *command.LogicalVolume
+        switch dc.Type {
+        case TypeThick:
+            lv, err = vg.CreateVolume(req.GetName(), requested, req.GetTags(), stripe, stripeSize, lvcreateOptions)
+        case TypeThin:
+            lv, err = pool.CreateVolume(req.GetName(), requested, req.GetTags(), stripe, stripeSize, lvcreateOptions)
+        default:
+            return nil, status.Error(codes.Internal, fmt.Sprintf("unsupported device class target: %s", dc.Type))
+        }
+        // ...
+        return &proto.CreateLVResponse{
+            Volume: &proto.LogicalVolume{
+                Name:     lv.Name(),
++/-             Size:     lv.Size(),
+                DevMajor: lv.MajorNumber(),
+                DevMinor: lv.MinorNumber(),
+            },
+        }, nil
+    }
+```
+
+### Dealing with less than 1Gi storage in the TopoLVM scheduler prioritization
+
+Currently, the scheduler uses the following scoring methodology for a node: `min(10, max(0, log2(capacity >> 30 / divisor)))`
+Note that the actual value of the capacity is stored in an annotation of the form `topolvm.GetCapacityKeyPrefix()+dc`,
+and _more importantly_, contains the value of the capacity in bytes.
+
+Our simplest suggestion here, in order to not break anything, is to clearly mark the existing capacity calculation as inaccurate and introduce
+a new lower bound for less than 1Gi storage in the method `capacityToScore`:
+
+```go
+    func capacityToScore(capacity uint64, divisor float64) int {
+        // avoid logarithm of zero, which diverges to negative infinity.
+        if capacity == 0 {
+            return 0
+        }
+    
++/-     var converted float64
++/-     if capacity > 1<<30 {
++/-         converted = math.Log2(float64(capacity) / divisor * float64(1<<30))
++/-     } else {
++/-         converted = 1
++/-     }
++/-     
++/-     switch {
++/-     case converted < 0:
++/-         return 0
++/-     case converted > 10:
++/-         return 10
++/-     default:
++/-         return int(converted)
++/-     }
+    }
+```
+
+This leads to every free Capacity between 1 byte and 1GB free to get the score of `1`, which means they will be scored the same.
+The only downside to this algorithm is that its precision is 1GB. So if there are 2 nodes with `500Mi` and `700Mi` free storage, both would receive the same score.
+
+The only way to get around this is by
+1. Changing the way existing scoring works to make room for byte-level comparison
+2. Creating a new scoring algorithm side-by-side and allow users to switch
+
+If we accept changing existing scoring under the assumption that all scored nodes will stay in the same order, just with higher precision we can just make use
+of the existing `item.Annotations[topolvm.GetCapacityKeyPrefix()+dc]` and make use of this byte amount (as `int64`) directly for scoring.
```

Comment:
The score value is limited in [0, 10]. This is why we use log2 to calculate the score value.
https://github.com/kubernetes/kube-scheduler/blob/master/extender/v1/types.go#L29

Reply:
While implementing this btw I noticed that node scoring is inaccurate for multiple deviceClasses as it only uses the smallest available device Class for the host priority. Is that a documented limitation?

Reply:
>a linear ratio conversion instead of having to rely on log2

Could you explain why you prefer a linear ratio?
We use log2 for scoring because the score range is much smaller than PVC size. Using log2, we can handle sizes from 1GB(2^0) to 1TB(2^10).
I assume you use MB-sized PVCs, log2 produces 0, so it is meaningless. Right?
On the other hand, using a linear ratio, we could not handle GB or TB PVC.
I have no clear idea how to solve this now, but it may be helpful to add nobs to change algorithms or upper limits.

>it only uses the smallest available device Class for the host priority.

The intention of this scoring is to select nodes with high capacity in order to leave room for resizing. The capacity of the smallest dc is therefore used to score the nodes.

Reply:
> Could you explain why you prefer a linear ratio?

So the advantage of a linear ratio is that we will always preserve the order of the nodes during scoring. The calculated scores may be different from log2, but we will always have the correct direction, no matter if bytes, gb, tb, etc. I tried to explain this in https://github.com/topolvm/topolvm/pull/754/files#diff-a5e76d9046c66bb6541fa6fa58911fdd82a13dfff9921cfa973e6826c78280cbR574-R580

> On the other hand, using a linear ratio, we could not handle GB or TB PVC.

I don't think this is the case. We can still handle GB or TB PVCs with linear ratio since it would work with byte precision. Linear algorithm guarantees to always preserve the correct order, which I think is what we want.

> The intention of this scoring is to select nodes with high capacity in order to leave room for resizing. The capacity of the smallest dc is therefore used to score the nodes.

This makes sense for one deviceClass. But if you have more than one deviceClass, then choosing the smallest deviceClass does not help the scheduling in the larger deviceClass right?

Reply:
I understand your algorithm now.
I could not find the algorithm before because github does not show diffs properly when doing force-push. So please don't use force-push.

>choosing the smallest deviceClass does not help the scheduling in the larger deviceClass right?
This is intentional in the multiple deviceClasses scenario too.
We don't know which deviceClass PVCs users will extend. At least other DCs are more expandable than the smallest DC, so choosing the smallest DC is a safe choice.

Reply:
> I understand your algorithm now. I could not find the algorithm before because github does not show diffs properly when doing force-push. So please don't use force-push.

Understood, thanks for bringing that up. I will do regular merges until we are finished on reviews and then squash stuff together.

> > choosing the smallest deviceClass does not help the scheduling in the larger deviceClass right?
> 
> This is intentional in the multiple deviceClasses scenario too. We don't know which deviceClass PVCs users will extend. At least other DCs are more expandable than the smallest DC, so choosing the smallest DC is a safe choice.

Maybe we are drifting off from the proposal here and I will open up a separate discussion. 

I believe I still don't see how using the smallest deviceClass is a safe choice. A deviceClass availableBytes is backed by a volumegroup, so just because one volumegroup has capacity does not mean the other volumegroup can be expanded to the same amount right? If you take the smallest available capacity from a volumegroup on one node, that does not mean that the other volume groups can be extended by the same amount, because the volume groups contain different devices.

I will resolve this conversation and then I will open up something separate for my question because I do not think it is related to the proposal changes.

Reply:
To be clear, do you expect to replace the original scoring algorithm with yours? Since the original algorithm uses log2 of absolute capacity as a score, using a minimum score ensures a minimum usable capacity. However, the ratio comparison makes no such assumption.

Reply:
Not sure if I understand your question correctly. The ratio comparison algorithm I showcased also has a case for minimum available capacity ([in this case it just checks >0](https://github.com/topolvm/topolvm/pull/754/files#diff-a5e76d9046c66bb6541fa6fa58911fdd82a13dfff9921cfa973e6826c78280cbR598-R602), but that can be adjusted). That way you will always score with 1 if there is a given minimum usable capacity.

I think we do not have to replace the old scoring algorithm but we should offer it alongside the other one. Otherwise we risk breaking legacy behavior.

I think we can simply switch algorithms via flag / env whatever we decide on.

Reply:
>I think we do not have to replace the old scoring algorithm but we should offer it alongside the other one. Otherwise we risk breaking legacy behavior.
>I think we can simply switch algorithms via flag / env whatever we decide on.

understood. thanks.

P.S. I left my concern, but it must not be solved now.
Assume that node A has a capacity of 1G for dc1 and 9G for dc2, and node B has a capacity of 4G for dc1 and 2G for dc2.
In this case, the old algorithm scores log(1) for node A and log(2) for node B, so node B wins.
On the other hand, your algorithm scores 1/4 for node A and 2/9 for node B, so node A wins.
The former ensures at least 2G usable capacity for both DCs, but the latter can only provide 1G for dc1.

Reply:
Interesting scenario, I just went and added these test cases into the test function below and i think the results are slightly different.

```go
func TestCapacityToScore(t *testing.T) {
	testCases := []struct {
		input   uint64
		divisor float64
		expect  int
	}{
		// NodeA - dc1=1Gi,dc2=9Gi
		// NodeB - dc1=4Gi,dc2=2Gi
		// dc1
		// NodeA
		{1 << 30, 1, 0},
		// NodeB
		{4 << 30, 1, 2}, // NodeB chosen for dc1
		// dc2
		// NodeA
		{9 << 30, 1, 3}, // NodeA chosen for dc2
		// dc2
		// NodeB
		{2 << 30, 1, 1},
		// NodeA: dc1=0,dc2=3, min-score=0
		// NodeB: dc1=2,dc2=1, min-score=2
		// Result => NodeB wins
	}

	for _, tt := range testCases {
		score := capacityToScore(tt.input, tt.divisor)
		if score != tt.expect {
			t.Errorf("score incorrect: input=%d expect=%d actual=%d",
				tt.input,
				tt.expect,
				score,
			)
		}
	}
}

func TestCapacityToScoreByRatio(t *testing.T) {
	testCases := []struct {
		input  string
		max    string
		expect int64
	}{
		// NodeA - dc1=1Gi,dc2=9Gi
		// NodeB - dc1=4Gi,dc2=2Gi
		// dc1-max=4Gi
		// dc2-max=9Gi
		//
		// dc1
		// NodeA
		{"1Gi", "4Gi", 2},
		// NodeB
		{"4Gi", "4Gi", 10}, // NodeB chosen for dc1
		// dc2
		// NodeA
		{"9Gi", "9Gi", 10}, // NodeA chosen for dc2
		// NodeB
		{"2Gi", "9Gi", 2},
		// NodeA: dc1=2,dc2=10, min-score=2
		// NodeB: dc1=10,dc2=2, min-score=2
		// Result => both nodes weighed equally, both nodes win equally
	}

	for i, tt := range testCases {
		t.Run(fmt.Sprintf("%v", i), func(t *testing.T) {
			in := resource.MustParse(tt.input)
			max := resource.MustParse(tt.max)
			score := capacityToScoreByRatio(in.Value(), max.Value())

			if score != tt.expect {
				t.Errorf("score incorrect: input=%v, max=%v expect=%v actual=%v",
					tt.input,
					tt.max,
					tt.expect,
					score,
				)
			}
		})

	}
}
```

As you can see in this attachment, the old algorithm indeed chooses node A as you said, but the proposed algorithm actually scores both nodes equally because both device classes have the same ratio of free space towards each other. meaning that pvcs will get distributed on both nodes. This is because the scoring algorithm actually chooses only the lowest scored dc for the entire node, even though technically this is inaccurate.

This is ultimately the issue i described above. Scoring via lowest scored device class per  node leaves out all other device classes for scoring. I believe this scoring is highly flawed because just because device-class1 has lots of capacity does not mean you can use it (e.g. for device-class2) without manual intervention

Reply:
> I also noticed that the pod capacity annotations are completely ignored right now, so if it expects 5 gb on dc1 and the node doesnt have that, its not reflected in the scoring.

Indeed. The annotations are only used for filtering. It may be better to take the request size into account. However, scoring is calculated one pod at a time, so overall optimisation may not be achieved. For this reason, we now only consider the capacity of the nodes.

Reply:
> It is because we use int precision. If I choose more appropriate values, A will win, IMO.

Of course if we have other ratios, the scores will change. Im not sure if this is related to int precision though. Could you elaborate?

> I described above, the old algorithm is to choose a node that has room to expand volumes even if you choose any DCs.

That does not make sense to me, sorry that Im taking so long to understand. How do you know if you can expand the volume by choosing A over B? The capacity is reporting volume group capacities, not device capacities. That means you cannot derive if you can expand based on this, right? Since you do not take into account the pod capacity, you dont know what is a better candidate.

That aside: we should definetly not replace the old algorithm for now and simply introduce a score of 1 for all capacities 1Gi>capacity>0 like suggested in option A. I believe we need a separate ideation for changing the scheduler.
 
I think if we introduce a second capacity scheduling algorithm besides the original one, we should be safe. Then we make the new one opt-in and for upgrades, everything will run the same for capacities above 1Gi. What do you think?

Reply:
>Of course if we have other ratios, the scores will change. Im not sure if this is related to int precision though. Could you elaborate?

I wanted to say that the values I gave as examples were not examples of A winning on the new algorithm when calculated with Int.
Here is a correct example A wins on the new algorithm but B wins on the old.
```
A: dc1: 1G, dc2: 7G
B: dc1: 2G, dc2: 3G
```

>That does not make sense to me, sorry that Im taking so long to understand.

Not your fault. I tried to explain but I couldn't do well.

>How do you know if you can expand the volume by choosing A over B? The capacity is reporting volume group capacities, not device capacities. That means you cannot derive if you can expand based on this, right? Since you do not take into account the pod capacity, you dont know what is a better candidate.

As you know, prioritization will run after filtering nodes, so we can choose any node from all possible nodes. And we can't know which volumes users will expand **in the future**. So we assume that all volumes will be extended. With such an assumption, we **think** it is better to choose a node that has the highest minimum usable capacity of DCs.

Consider my example, both node A and node B are candidates, the volume using dc1 is at most 1G and another volume using dc2 is at most 3G because filtering is already done. In this case, the better choice is B because A does not provide more capacity for dc1 if a volume using dc1 is 1G. Of course, if the volume using dc2 is 3G and a user wants to expand the volume, A should be chosen.
As you can see, there is no best algorithm, but the old one works well in our experience and we do not want to break it.

>That aside: we should definetly not replace the old algorithm for now and simply introduce a score of 1 for all capacities 1Gi>capacity>0 like suggested in option A. I believe we need a separate ideation for changing the scheduler.
>
>I think if we introduce a second capacity scheduling algorithm besides the original one, we should be safe. Then we make the new one opt-in and for upgrades, everything will run the same for capacities above 1Gi. What do you think?

It is awesome. I approve of the idea.

Reply:
Thanks so much for explaining, now it actually makes so much more sense to me!! And you are completely right, in this case the new algorithm proposal does not achieve our goal and I believe we should stick with the old one.

I now removed the new algorithm proposal with the note to create a separate enhancement proposal on how to enhance the scheduler precision.

I think with this we have the last big question of the proposal answered: We will simply stick to Gi level precision in the scheduler, and live with the fact that all capacities <1Gi are scored the same.

Reply:
Understood. Thanks, too.